### ISSUE TITLE
[@typescript-eslint/no-use-before-define] False positive while eslint no-use-before-define is either disabled or no issue

### ISSUE THREAD

**Issue by blackxored** [typescript-eslint/typescript-eslint#1064]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-use-before-define": ["error"]
  }
}
```

```TS

const updateWidgetSettings = (_) => {
  return {
    effects: [Http(updateSettings)]
  }
}

const updateSettings = () => {};
```

**Expected Result**
No errors
**Actual Result**
'updateSettings` was used before it was defined.

**Additional Info**
I've never had issue with this kind of code, either I've disabled the eslint rule in the past or have never been an issue, as I'd like to put helper/pseudoprivate methods at the bottom of my files.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.3` |
| `@typescript-eslint/parser`        | `2.3.3` |
| `TypeScript`                       | `3.5.2` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `12.11.0` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
This is a valid warning.

ESLint does not do complex control-flow analysis as part of the rule, so it cannot tell that you don't call `updateWidgetSettings` before you define `updateSettings`.

I.e. if you were to insert a line like so, your code would break:
```ts
const updateWidgetSettings = (_) => {
  return {
    effects: [Http(updateSettings)]
  }
}

updateWidgetSettings(); // error - Cannot access 'updateSettings' before initialization

const updateSettings = () => {};
```

Sure your example is simple, but there could easily be 100 lines between the two variable declarations, and if dev doesn't realise things are ordered this way, then you will get a bug.

Because ESLint doesn't do complex control-flow analysis - it errs on the side of caution, warning you of a _potential_ (and probable) problem.
